### PR TITLE
MNT: protect from out-of-bounds data access at the c level

### PR DESCRIPTION
**Pull request by tacaswell**
As suggested by @cgohlke

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_tk.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    fig, ax = plt.subplots()
+    for bad_boxes in ((-1, 2, 0, 2),
+                      (2, 0, 0, 2),
+                      (1, 6, 0, 2),
+                      (0, 2, -1, 2),
+                      (0, 2, 2, 0),
+                      (0, 2, 1, 6),
+    ):
+        with pytest.raises(ValueError):
+            print(bad_boxes)
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_tk.py]: remove the print?

File: src/_tkagg.cpp [hunk 2]
```diff
@@ -67,6 +67,12 @@ static PyObject *mpl_tk_blit(PyObject *self, PyObject *args)
         PyErr_SetString(PyExc_ValueError, "Failed to extract Tk_PhotoHandle");
         goto exit;
     }
+    if (0 > y1 || y1 > y2 || y2 > height ||
+        0 > x1 || x1 > x2 || x2 > width ) {
```
Comment (anntzer) [inline src/_tkagg.cpp]: extra space after width?

**Comment by tacaswell** [comment]
I still can write code I swear...

**Comment by tacaswell** [comment]
Can probably re-use the `skip_on_importfailure` flag to further simplify the Qt tests?

**Comment by anntzer** [comment]
quite possibly, indeed, looks like a good idea.

File: src/_tkagg.cpp [hunk 3]
```diff
@@ -67,6 +67,11 @@ static PyObject *mpl_tk_blit(PyObject *self, PyObject *args)
         PyErr_SetString(PyExc_ValueError, "Failed to extract Tk_PhotoHandle");
         goto exit;
     }
+    if (0 > y1 || y1 > y2 || y2 > height || 0 > x1 || x1 > x2 || x2 > width) {
```
Comment (timhoffm) [inline src/_tkagg.cpp]:
```suggestion
    if (y1 < 0 || y1 > y2 || y2 > height || x1 < 0 || x1 > x2 || x2 > width) {
```
I find it easier this way "y1 is smaller than 0 or larger than y2". `0 > y1` sort of ties a knot in my brain. :smile:

Comment (tacaswell) [reply src/_tkagg.cpp]: That is fair, but having all of the `>` go the same way is also helpfule...

Moot now that @efiring merged it..

**Pull request merged**